metric-example/query-by-example.md

```diff
@@ -0,0 +1,516 @@
+```
+builtin:host.cpu.usage:avg  
+builtin:host.cpu.usage:fold:max
+```
+The second line merges the statistical data over all time slots, and then extracts the maximum value from the statistical data. This works, but the folded value will be actual maximum sample. If we draw the series and the maximum together into a chart, the folded value may exceed the highest average value, since each value in the chart is actually an aggregate of multiple samples (the average of samples in our case). What if we want to query for the maximum of average values in the chart, not the actual maximum sample?
+
+For that use case, we want to build a list of averages rather than a list of original samples, and extract the maximum from these averages.
+
+We can take advantage of the fact that `:splitBy` collects pre-aggregated input values (e.g. averages) into a list if the next aggregation after the merge requires it. If we pass all dimensions of the metric to `:splitBy,` no dimensions will be removed and hence no actual merging will happen, instead we have a list with a single average for each time slots. Before the final aggregation, we need to insert `:fold` so that the lists are merged into one over time. From this merged list, we finally extract the maximum with `:max`. Taking all of this together we end up with this selector:
+```
+builtin:host.cpu.usage:avg:splitBy(dt.entity.host):fold:max
+```
+The result is exactly what we wanted: the maximum of average values over time - We can draw the maximum into the chart and see that it will exactly touch the peak of the graph.
+
+If you query this selector together with `builtin:host.cpu.usage:avg` you see that the single value is always exactly equal to the maximum entry of the corresponding series.
+
+# Scenario 15: Filtering for Special Characters with Quoting
+
+>**Task**
+>We have a filter text field, but the resulting metric selector is invalid if we type in special characters.
+
+Some characters have a special meaning in metric selectors: ( ) : , ~ "
+
+Additionally, white space is used to tell where an identifier such as "dt.entity.host" ends. Replacing the periods (full stops) with spaces would result in three identifiers next to each other, rather than a single one that contains white space.
+
+If white space or any of mentioned characters are to be used within a dimension name, in a filter matcher reference, or otherwise in their "literal" sense, rather than as syntax, we need to be careful and escape or quote them. Knowing about proper escaping is especially important when we build our selectors in code that handles user input. In such scenarios, e.g. when a user can provide a filter reference in a text field, we want to prevent them from making the resulting selector syntactically invalid or even worse, allow them to change its meaning and insert a malicious piece of code into our selector.

```

Comment:
```suggestion
If white space or any of the characters mentioned above are to be used within a dimension name, a filter matcher reference, or otherwise in their "literal" sense, rather than as syntax, we need to escape or quote them. Knowing about proper escaping is especially important when we build our selectors in code that handles user input (e.g. when a user can provide a filter reference in a text field). In such scenarios, we want to prevent them from making the resulting selector syntactically invalid or, even worse, allow them to change its meaning and insert a malicious piece of code into our selector.

```